src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder.cs

```diff
@@ -25,573 +25,649 @@ namespace Microsoft.CodeAnalysis.FindSymbols
+            // First see what derived metadata types we might find in this project.
+            // This is only necessary if we started with a metadata type.
+            if (searchInMetadata)
             {
-                List<INamedTypeSymbol> allTypes;
-                if (locationsInMetadata)
-                {
-                    // From metadata, have to check other (non private) metadata types, as well as
-                    // source types.
-                    allTypes = GetAllSourceAndAccessibleTypesInCompilation(compilation, cancellationToken);
-                }
-                else
-                {
-                    // It's from source, so only other source types could derive from it.
-                    allTypes = GetAllSourceTypesInCompilation(compilation, cancellationToken);
-                }
+                var foundMetadataTypes = await FindMetadataTypesInProjectAsync(
+                    currentMetadataTypes, project, metadataTypeMatches, cancellationToken).ConfigureAwait(false);
 
-                dependentTypeIds = new List<SymbolKey>();
-                foreach (var t in allTypes)
+                foreach (var foundType in foundMetadataTypes)
                 {
-                    cancellationToken.ThrowIfCancellationRequested();
+                    Debug.Assert(foundType.Locations.Any(s_isInMetadata));
 
-                    if (predicate(t, type))
+                    // Add to the result list.
+                    result.Add(foundType);
+
+                    if (transitive && shouldContinueSearching(foundType))
                     {
-                        dependentTypeIds.Add(t.GetSymbolKey());
+                        currentMetadataTypes.Add(foundType);
+                        currentSourceAndMetadataTypes.Add(foundType);
                     }
                 }
-
-                dependentTypeIds = GetOrAddDependentTypes(cache, compilation, typeId, dependentTypeIds);
             }
 
-            foreach (var id in dependentTypeIds)
+            // Now search the project and see what source types we can find.
+            var foundSourceTypes = await FindSourceTypesInProjectAsync(
+                currentSourceAndMetadataTypes, project, 
+                sourceTypeImmediatelyMatches,
+                shouldContinueSearching,
+                transitive, cancellationToken).ConfigureAwait(false);
+
+            foreach (var foundType in foundSourceTypes)
             {
-                cancellationToken.ThrowIfCancellationRequested();
+                Debug.Assert(foundType.Locations.All(s_isInSource));
 
-                var resolvedSymbols = id.Resolve(compilation, cancellationToken: cancellationToken).GetAllSymbols();
-                foreach (var resolvedSymbol in resolvedSymbols)
+                // Add to the result list.
+                result.Add(foundType);
+
+                if (transitive && shouldContinueSearching(foundType))
```

Comment:
Why does shouldContinueSearching cause us to conditionally add it to the result HashSets? I would have imagined by the name it meant we'd just bail the loop.